specs/src/shwap/spec.md

```diff
@@ -0,0 +1,267 @@
+pluggability to rapidly scaffold Celestia's DA network. However, it came with the price of scalability limits and
+round-trips resulting in BS slower than block production. Before the network launch, the transition
+to the optimized [ShrEx protocol][shrex] for BS and integrating [CAR and DAGStore-based storage][storage] happened
+optimizing BS and ND. However, DAS was left untouched, preserving its weak scalability and roundtrip inefficiency. Shwap
+addresses these and provides an extensible and flexible framework for BS, ND, and beyond.
+
+## Terms and Definitions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in BCP
+14 [RFC2119] [RFC8174] when, and only when, they appear in all
+capitals, as shown here.
+
+Commonly used terms in this document are described below.
+
+_**Shwap**_: The protocol described by this document. Shwap is a
+portmanteau name of words share and swap.
+
+_**[Share][shares]**_: The core data structure of DataSquare **"swapped"** between peers.
+
+_**[DataSquare][square]**_: The DA square format used by Celestia DA network.
+
+_**[DAH][dah]**_: The Data Availability Header with Row and Column commitments.
+
+_**[Namespace][ns]**_: The namespace grouping sets of shares.
+
+_**Peer**_: An entity that can participate in a Shwap protocol. There are three types of peers:
+client, server and node.
+
+_**Client**_: The Peer that requests content by content identifies over Shwap.
+
+_**Server**_: The Peer that responds with content over Shwap.
+
+_**Node**_: The peer that namespacesis both the client and the server.
+
+_**Proof**_: Merkle inclusion proof of the data in the DataSquare.
+
+## Rationale
+
+### Multihashes and CID
+
+Shwap takes inspiration from content addressability, but breaks-free from hash-based only model to optimize message sizes
+and data request patterns. In some way, it hacks into multihash abstraction to make it contain data that isn't in fact a
+hash. Furthermore, the protocol does not include hash digests in the multihashes. The authentication of the messages
+happens using externally provided data commitment.
+
+## Protocol Dependencies
+
+### Bitswap
+
+Shwap depends on Bitswap for swapping bits in fully distributed p2p-manner.
+
+## Share Identifiers
+
+This section defines list of supported share identifiers. Share identifiers defined by Shwap can be used to uniquely
+identify any [share container](#share-containers) over a chain with arbitrary number of [DataSquares][square], like a range of 
+[shares][shares], a row or a [blob][blob]. Every share identifier relates to a respective share container and wise-versa.
+
+Identifiers are embeddable to narrow down to the needed content. (TODO: Describe better)
+
+Identifiers MUST have a fixed size for their fields. Subsequently, protobuf can't be used for CID serialization due to
```

Comment:
That's a good point. I really would like not to have bitswap as a rationale for this, as this breaks Shwap protocol independence. Ideally, we would find a way to avoid this limitation until the CIP is finalized or find another rationale for why this is necessary.